vending-app/src/main/kotlin/com/google/android/finsky/splitinstallservice/extensions.kt

```diff
@@ -0,0 +1,347 @@
+    val response = httpClient.requestDownloadUrl(context, requestUrl.toString(), oauthToken, langName.toList())
+    Log.d(TAG, "getDownloadUrls: requestDownloadUrl end response -> $response")
+    val splitPkgInfoList = response?.response?.splitReqResult?.pkgList?.pkgDownLoadInfo ?: throw RuntimeException("splitPkgInfoList is null")
+    val downloadUrls = ArrayList<Array<String>>()
+    splitPkgInfoList.filter { !it.splitPkgName.isNullOrEmpty() && !it.downloadUrl.isNullOrEmpty() }.forEach { info ->
+        langName.filter { "config.$it".contains(info.splitPkgName!!) }.forEach { downloadUrls.add(arrayOf(it, info.downloadUrl!!)) }
+        splitName.filter { it.contains(info.splitPkgName!!) }.forEach { downloadUrls.add(arrayOf(it, info.downloadUrl!!)) }
+    }
+    return downloadUrls
+}
+
+private suspend fun HttpClient.requestDownloadUrl(context: Context, requestUrl: String, auth: String, requestLanguagePackage: List<String>) = runCatching {
+    val androidId = SettingsContract.getSettings(
+        context, SettingsContract.CheckIn.getContentUri(context), arrayOf(SettingsContract.CheckIn.ANDROID_ID)
+    ) { cursor: Cursor -> cursor.getLong(0) }
+    Log.d(TAG, "requestUrl->$requestUrl")
+    Log.d(TAG, "auth->$auth")
+    Log.d(TAG, "androidId->$androidId")
+    Log.d(TAG, "requestLanguagePackage->$requestLanguagePackage")
+    get(url = requestUrl, headers = getLicenseRequestHeaders(auth, 1).toMutableMap().apply {
+        val xPsRh = String(
+            Base64.encode(
+                getDefaultLicenseRequestHeaderBuilder(1).languages(RequestLanguagePackage.Builder().language(requestLanguagePackage).build()).build().encode().encodeGzip(),
+                Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING
+            )
+        )
+        put("X-PS-RH", xPsRh)
+    }.onEach {
+        Log.d(TAG, "key:${it.key}  value:${it.value}")
+    }, adapter = GoogleApiResponse.ADAPTER)
+}.onFailure {
+    Log.d(TAG, "requestDownloadUrl: ", it)
+}.getOrNull()
+
+private suspend fun HttpClient.downloadSplitPackage(context: Context, downloadUrls: ArrayList<Array<String>>): Boolean = coroutineScope {
+    val results = downloadUrls.map { urls ->
+        Log.d(TAG, "downloadSplitPackage: ${urls.contentToString()}")
+        async {
+            runCatching {
+                download(urls[1], File(context.splitSaveFile().toString(), urls[0]), SPLIT_INSTALL_REQUEST_TAG)
+            }.onFailure {
+                Log.w(TAG, "downloadSplitPackage urls:${urls.contentToString()}: ", it)
+            }.getOrNull() != null
+        }
+    }.awaitAll()
+    return@coroutineScope results.all { it }
+}
+
+private suspend fun installSplitPackage(context: Context, httpClient: HttpClient, downloadUrl: ArrayList<Array<String>>, packageName: String, language: String?): Intent {
+    Log.d(TAG, "installSplitPackage downloadUrl: ${downloadUrl.firstOrNull()}")
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
```

Comment:
This check is only executed after calling `getDownloadUrls`. I would suggest performing the check earlier in the flow, and annotating relevant methods with `@RequiresApi(Build.VERSION_CODES.LOLLIPOP)`.